<a href="https://colab.research.google.com/github/Mirucla/Applied-Data-Management/blob/main/ML_HW_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# PART THREE

In [2]:
from sklearn.datasets import fetch_california_housing

ca_house_db = fetch_california_housing()

print(ca_house_db.data.shape)
print(ca_house_db.target.shape)
print(ca_house_db.feature_names)
print(ca_house_db.DESCR)
print(ca_house_db)

(20640, 8)
(20640,)
['MedInc', 'HouseAge', 'AveRooms', 'AveBedrms', 'Population', 'AveOccup', 'Latitude', 'Longitude']
.. _california_housing_dataset:

California Housing dataset
--------------------------

**Data Set Characteristics:**

    :Number of Instances: 20640

    :Number of Attributes: 8 numeric, predictive attributes and the target

    :Attribute Information:
        - MedInc        median income in block group
        - HouseAge      median house age in block group
        - AveRooms      average number of rooms per household
        - AveBedrms     average number of bedrooms per household
        - Population    block group population
        - AveOccup      average number of household members
        - Latitude      block group latitude
        - Longitude     block group longitude

    :Missing Attribute Values: None

This dataset was obtained from the StatLib repository.
https://www.dcc.fc.up.pt/~ltorgo/Regression/cal_housing.html

The target variable is the median ho

In [3]:
import numpy as np

print(np.isnan(ca_house_db.data).any())
print(np.isinf(ca_house_db.data).any())
print(np.isnan(ca_house_db.target).any())
print(np.isinf(ca_house_db.target).any())


False
False
False
False


# Question 1

In [4]:
from sklearn.model_selection import train_test_split

X_train1, X_test, y_train1, y_test = train_test_split(ca_house_db.data, ca_house_db.target, test_size=0.5,
                                                      random_state=42)
X_train2, X_val, y_train2, y_val = train_test_split(X_train1, y_train1, test_size=0.2, random_state=42)

print(X_train1.shape, y_train1.shape)
print(X_train2.shape, y_train2.shape)
print(X_val.shape, y_val.shape)
print(X_test.shape, y_test.shape)

(10320, 8) (10320,)
(8256, 8) (8256,)
(2064, 8) (2064,)
(10320, 8) (10320,)


# Question 2

In [5]:
pip install tensorflow

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [6]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

def create_model(num_hidden_layers, num_units, optimizer, activation, output_function):
    # Create a new sequential model
    model = Sequential()
    # Add the input layer with the specified number of units and activation function
    model.add(Dense(num_units, input_shape=(X_train2.shape[1],), activation=activation))
    # Add the specified number of hidden layers with the specified number of units and activation function
    for i in range(num_hidden_layers):
        model.add(Dense(num_units, activation=activation))
    # Add the output layer with a single unit and the specified output activation function
    model.add(Dense(1, activation=output_function))
    # Compile the model with the mean squared error loss, the specified optimizer, and mean squared error metric
    model.compile(loss='mean_squared_error', optimizer=optimizer, metrics=['mse'])
    # Return the compiled model
    return model

In [7]:
import numpy as np

# set of hyperparameters 1
num_hidden_layers_1 = 2
num_units_1 = 32
optimizer_1 = 'adam'
activation_1 = 'tanh'
output_function_1 = 'linear'

model_1 = create_model(num_hidden_layers_1, num_units_1, optimizer_1, activation_1, output_function_1)
history_1 = model_1.fit(X_train2, y_train2, epochs=100, batch_size=32, validation_data=(X_val, y_val))

# set of hyperparameters 2
num_hidden_layers_2 = 3
num_units_2 = 64
optimizer_2 = 'adam'
activation_2 = 'sigmoid'
output_function_2 = 'linear'

model_2 = create_model(num_hidden_layers_2, num_units_2, optimizer_2, activation_2, output_function_2)
history_2 = model_2.fit(X_train2, y_train2, epochs=100, batch_size=32, validation_data=(X_val, y_val))

# set of hyperparameters 3
num_hidden_layers_3 = 1
num_units_3 = 16
optimizer_3 = 'adam'
activation_3 = 'relu'
output_function_3 = 'linear'

model_3 = create_model(num_hidden_layers_3, num_units_3, optimizer_3, activation_3, output_function_3)
history_3 = model_3.fit(X_train2, y_train2, epochs=100, batch_size=32, validation_data=(X_val, y_val))

Epoch 1/100
258/258 [==============================] - 3s 6ms/step - loss: 1.3724 - mse: 1.3724 - val_loss: 1.4161 - val_mse: 1.4161
Epoch 2/100
258/258 [==============================] - 1s 5ms/step - loss: 1.3336 - mse: 1.3336 - val_loss: 1.3859 - val_mse: 1.3859
Epoch 3/100
258/258 [==============================] - 2s 6ms/step - loss: 1.3337 - mse: 1.3337 - val_loss: 1.3836 - val_mse: 1.3836
Epoch 4/100
258/258 [==============================] - 2s 6ms/step - loss: 1.3336 - mse: 1.3336 - val_loss: 1.3818 - val_mse: 1.3818
Epoch 5/100
258/258 [==============================] - 1s 5ms/step - loss: 1.3299 - mse: 1.3299 - val_loss: 1.3842 - val_mse: 1.3842
Epoch 6/100
258/258 [==============================] - 2s 7ms/step - loss: 1.3297 - mse: 1.3297 - val_loss: 1.3827 - val_mse: 1.3827
Epoch 7/100
258/258 [==============================] - 2s 7ms/step - loss: 1.3280 - mse: 1.3280 - val_loss: 1.3891 - val_mse: 1.3891
Epoch 8/100
258/258 [==============================] - 2s 8ms/step - 

In [8]:
X_test

array([[   1.6812    ,   25.        ,    4.19220056, ...,    3.87743733,
          36.06      , -119.01      ],
       [   2.5313    ,   30.        ,    5.03938356, ...,    2.67979452,
          35.14      , -119.46      ],
       [   3.4801    ,   52.        ,    3.97715472, ...,    1.36033229,
          37.8       , -122.44      ],
       ...,
       [   2.5208    ,   52.        ,    3.90816327, ...,    2.28571429,
          37.4       , -122.08      ],
       [   2.7461    ,   17.        ,    5.42619048, ...,    2.85714286,
          36.15      , -120.38      ],
       [   1.8875    ,   34.        ,    4.24712644, ...,    4.98563218,
          34.02      , -118.19      ]])

In [9]:
from sklearn.metrics import r2_score

# Evaluate model 1
y_pred_1 = model_1.predict(X_val)
r2_1 = r2_score(y_val, y_pred_1)
print("R2 score for model 1: {:.3f}".format(r2_1))

# Evaluate model 2
y_pred_2 = model_2.predict(X_val)
r2_2 = r2_score(y_val, y_pred_2)
print("R2 score for model 2: {:.3f}".format(r2_2))

# Evaluate model 3
y_pred_3 = model_3.predict(X_val)
r2_3 = r2_score(y_val, y_pred_3)
print("R2 score for model 3: {:.3f}".format(r2_3))

65/65 [==============================] - 0s 2ms/step
R2 score for model 1: 0.350
65/65 [==============================] - 0s 2ms/step
R2 score for model 2: 0.635
65/65 [==============================] - 0s 2ms/step
R2 score for model 3: 0.439


# Therefore, the model with high R2 is Model 2. It's R2 score is 63.5% that is way much better than Model 1 and Model 3. 

# Question 3

In [13]:
y_pred_test = model_2.predict(X_test)

# Calculate R2 score on testing data
r2_test = r2_score(y_test, y_pred_test)

# Print R2 score rounded to 2 decimal places
print("R2 score on testing data:", round(r2_test, 3))

323/323 [==============================] - 1s 2ms/step
R2 score on testing data: 0.638


This model is consistent with X_test data as well. That the Model 2 has the same R2 score on all the testing data. 

# Question 4

In [11]:
# Calculate the absolute error for each testing sample
abs_errors = np.abs(y_pred_test - y_test)

# Sort the samples by absolute error
sorted_indices = np.argsort(abs_errors, axis=0)[::-1]

# Print the ten samples with the largest absolute errors
for i in range(10):
    index = sorted_indices[i]
    print(f'Sample {index}: predicted value = {y_pred_test[index]}, actual value = {y_test[index]}, absolute error = {abs_errors[index]}')

Sample [4806 4806 2090 ... 2090 4806 4806]: predicted value = [[4.7995405 ]
 [4.7995405 ]
 [0.66341794]
 ...
 [0.66341794]
 [4.7995405 ]
 [4.7995405 ]], actual value = [5.00001 5.00001 1.375   ... 1.375   5.00001 5.00001], absolute error = [[4.32254052 4.34154052 0.20046948 ... 1.63254052 4.09954052 3.43254052]
 [4.32254052 4.34154052 0.20046948 ... 1.63254052 4.09954052 3.43254052]
 [0.18641794 0.20541794 4.33659206 ... 2.50358206 0.03658206 0.70358206]
 ...
 [0.18641794 0.20541794 4.33659206 ... 2.50358206 0.03658206 0.70358206]
 [4.32254052 4.34154052 0.20046948 ... 1.63254052 4.09954052 3.43254052]
 [4.32254052 4.34154052 0.20046948 ... 1.63254052 4.09954052 3.43254052]]
Sample [5925 5925 6741 ... 6741 5925 5925]: predicted value = [[4.730632 ]
 [4.730632 ]
 [0.6959078]
 ...
 [0.6959078]
 [4.730632 ]
 [4.730632 ]], actual value = [4.656 4.656 1.875 ... 1.875 4.656 4.656], absolute error = [[4.25363183e+00 4.27263183e+00 2.69378172e-01 ... 1.56363183e+00
  4.03063183e+00 3.36363183e

# Source codes: 1) Class Lecture , 2) Youtube videos 3) ChatGPT 